bench-tps/src/bench_tps_client/high_tps_client.rs

```diff
@@ -0,0 +1,233 @@
+//! This is achieved by:
+//! * optimizing the size of send batch
+//! * sending batches using connection cache in nonblocking fashion
+//! * use leader cache from TpuClient to minimize rpc calls
+//! * using several staked connection caches (not implemented yet)
+//!
+use {
+    crate::bench_tps_client::{BenchTpsClient, BenchTpsError, Result},
+    solana_client::{
+        connection_cache::Protocol, nonblocking::tpu_client::LeaderTpuService,
+        tpu_connection::TpuConnection,
+    },
+    solana_connection_cache::connection_cache::ConnectionCache as BackendConnectionCache,
+    solana_quic_client::{QuicConfig, QuicConnectionManager, QuicPool},
+    solana_rpc_client::rpc_client::RpcClient,
+    solana_rpc_client_api::config::RpcBlockConfig,
+    solana_sdk::{
+        account::Account, commitment_config::CommitmentConfig, epoch_info::EpochInfo, hash::Hash,
+        message::Message, pubkey::Pubkey, signature::Signature, slot_history::Slot,
+        transaction::Transaction,
+    },
+    solana_transaction_status::UiConfirmedBlock,
+    std::sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
+
+#[derive(Debug, Clone, Copy)]
+pub struct HighTpsClientConfig {
+    pub send_batch_size: usize,
+    pub fanout_slots: u64,
+}
+
+type ConnectionCache = BackendConnectionCache<QuicPool, QuicConnectionManager, QuicConfig>;
+pub struct HighTpsClient {
+    leader_tpu_service: LeaderTpuService,
+    exit: Arc<AtomicBool>,
+    rpc_client: Arc<RpcClient>,
+    connection_cache: Arc<ConnectionCache>,
+    config: HighTpsClientConfig,
+}
+
+impl Drop for HighTpsClient {
+    fn drop(&mut self) {
+        self.exit.store(true, Ordering::Relaxed);
+    }
+}
+
+impl HighTpsClient {
+    pub fn new(
+        rpc_client: Arc<RpcClient>,
+        websocket_url: &str,
+        config: HighTpsClientConfig,
+        connection_cache: Arc<ConnectionCache>,
+    ) -> Result<Self> {
+        let exit = Arc::new(AtomicBool::new(false));
+        let create_leader_tpu_service = LeaderTpuService::new(
+            rpc_client.get_inner_client().clone(),
+            websocket_url,
+            Protocol::QUIC,
+            exit.clone(),
+        );
+        let leader_tpu_service = tokio::task::block_in_place(|| {
+            rpc_client.runtime().block_on(create_leader_tpu_service)
+        })?;
+        Ok(Self {
+            leader_tpu_service,
+            exit,
+            rpc_client,
+            connection_cache,
+            config,
+        })
+    }
+}
+
+impl BenchTpsClient for HighTpsClient {
+    fn send_transaction(&self, transaction: Transaction) -> Result<Signature> {
+        self.rpc_client
+            .send_transaction(&transaction)
+            .map_err(|err| err.into())
+    }
+
+    fn send_batch(&self, transactions: Vec<Transaction>) -> Result<()> {
+        let wire_transactions = transactions
+            .into_iter() //.into_par_iter() any effect of this?
```

Comment:
in response to comment here: probably if `Vec<Transaction>` is long. `tpu_client` uses `.into_par_iter()` so may make sense to use `into_par_iter()` but i haven't tested it to see performance diff